crypto/machine.go

```diff
@@ -180,15 +190,192 @@ func (mach *OlmMachine) HandleToDeviceEvent(evt *event.Event) {
+			if err != nil {
+				mach.Log.Error("Error exporting key for session %v: %v", igs.ID(), err)
+				return
+			}
+
+			forwardedRoomKey := event.Content{
+				Parsed: &event.ForwardedRoomKeyEventContent{
+					RoomKeyEventContent: event.RoomKeyEventContent{
+						Algorithm:  id.AlgorithmMegolmV1,
+						RoomID:     igs.RoomID,
+						SessionID:  igs.ID(),
+						SessionKey: exportedKey,
+					},
+					SenderKey:          content.Body.SenderKey,
+					ForwardingKeyChain: igs.ForwardingChains,
+					SenderClaimedKey:   igs.SigningKey,
+				},
+			}
+
+			if err := mach.SendEncryptedToDevice(evt.Sender, content.RequestingDeviceID, forwardedRoomKey); err != nil {
+				mach.Log.Error("Failed to send encrypted forwarded key: %v", err)
+			}
+
+			mach.Log.Debug("Sent encrypted forwarded key to device %v for session %v", content.RequestingDeviceID, igs.ID())
+		}
+	// TODO add m.verification.start case
 	default:
 		deviceID, _ := evt.Content.Raw["device_id"].(string)
 		mach.Log.Trace("Unhandled to-device event of type %s from %s/%s", evt.Type.Type, evt.Sender, deviceID)
 	}
 }
 
+// SendEncryptedToDevice sends an Olm-encrypted event to the given user device.
+func (mach *OlmMachine) SendEncryptedToDevice(userID id.UserID, deviceID id.DeviceID, content event.Content) error {
+	// get device identity
+	device, err := mach.CryptoStore.GetDevice(userID, deviceID)
+	if err != nil {
+		return err
+	} else if device == nil {
+		// fetch if not found
+		usersToDevices := mach.fetchKeys([]id.UserID{userID}, "", true)
+		if devices, ok := usersToDevices[userID]; ok {
+			if device, ok = devices[deviceID]; !ok {
+				return errors.Errorf("Failed to get identity for device %v", deviceID)
+			}
+		} else {
+			return errors.Errorf("Error fetching devices for user %v", userID)
+		}
+	}
+	// create outbound sessions if missing
+	if err := mach.createOutboundSessions(map[id.UserID]map[id.DeviceID]*DeviceIdentity{
+		userID: {
+			deviceID: device,
+		},
+	}); err != nil {
+		return err
+	}
+
+	// get Olm session
+	olmSess, err := mach.CryptoStore.GetLatestSession(device.IdentityKey)
+	if err != nil {
+		return err
+	}
+	if olmSess == nil {
+		return err
+	}
+
+	encrypted := mach.encryptOlmEvent(olmSess, device, event.ToDeviceForwardedRoomKey, content)
+	encryptedContent := &event.Content{Parsed: &encrypted}
+
+	_, err = mach.Client.SendToDevice(event.ToDeviceEncrypted,
+		&mautrix.ReqSendToDevice{
+			Messages: map[id.UserID]map[id.DeviceID]*event.Content{
+				userID: {
+					deviceID: encryptedContent,
+				},
+			},
+		},
+	)
+
+	return err
+}
+
+// RequestRoomKey sends a key request for a room to the current user's devices. If the context is cancelled, then so is the key request.
+// Returns a bool channel that will get notified either when the key is received or the request is cancelled.
+func (mach *OlmMachine) RequestRoomKey(ctx context.Context, toUser id.UserID, toDevice id.DeviceID,
```

Comment:
It would probably be a good idea to split `RequestRoomKey`, `importForwardedRoomKey` and maybe the `*event.RoomKeyRequestEventContent` handling to a separate `keysharing.go` file. I'm not completely sure where `SendEncryptedToDevice` should be

Reply:
Good idea, moved. I kept `SendEncryptedToDevice` there however as it's supposed to be a generic utility function.